### PR TITLE
fix: Comprehensive CSRF vulnerability fix for CVE-2024-43338

### PR DESCRIPTION
**Pull request by ovidiul**
## Summary

This PR provides a comprehensive fix for the CSRF vulnerability (CVE-2024-43338), implementing multi-layered security measures to prevent unauthorized access to media upload functionality. **This PR supersedes and replaces PR #110.**

## Security Issues Addressed

- **CVE-2024-43338**: CSRF vulnerability in polls media upload functionality
- **Unauthorized Filter Registration**: Prevents malicious registration of media upload filters
- **Cross-User Attack Prevention**: User-specific nonces prevent cross-user CSRF attacks

### Frontend Security (js/polldaddy.js) - **MINIMAL CHANGES ONLY**

**🎯 Key Point**: The original minified structure is preserved. Only 3 vulnerable URL constructions were modified.

#### **Before (Vulnerable):**
```javascript
// Image upload - VULNERABLE
tb_show("Add an Image","media-upload.php?type=image&polls_media=1TB_iframe=1");

// Video upload - VULNERABLE  
tb_show("Add Video","media-upload.php?type=video&tab=type_url&polls_media=1&TB_iframe=1");

// Audio upload - VULNERABLE
tb_show("Add Audio","media-upload.php?type=audio&polls_media=1&TB_iframe=1");
```

#### **After (Secure):**
```javascript
// Image upload - SECURE
var r="media-upload.php?type=image&polls_media=1";
if(typeof pollsMediaSecurity!=="undefined"&&pollsMediaSecurity.nonce){
  r+="&_wpnonce="+encodeURIComponent(pollsMediaSecurity.nonce)
}
tb_show("Add an Image",r+"&TB_iframe=1");

// Video upload - SECURE
var r="media-upload.php?type=video&tab=type_url&polls_media=1";
if(typeof pollsMediaSecurity!=="undefined"&&pollsMediaSecurity.nonce){
  r+="&_wpnonce="+encodeURIComponent(pollsMediaSecurity.nonce)
}
tb_show("Add Video",r+"&TB_iframe=1");

// Audio upload - SECURE
var r="media-upload.php?type=audio&polls_media=1";
if(typeof pollsMediaSecurity!=="undefined"&&pollsMediaSecurity.nonce){
  r+="&_wpnonce="+encodeURIComponent(pollsMediaSecurity.nonce)
}
tb_show("Add Audio",r+"&TB_iframe=1");
```

**JavaScript Security Features:**
- ✅ **Minimal Impact**: Only changed the 3 vulnerable URL constructions
- ✅ **Structure Preserved**: Original minified code structure maintained
- ✅ **Graceful Degradation**: Works if `pollsMediaSecurity.nonce` is unavailable
- ✅ **Proper Encoding**: Uses `encodeURIComponent()` for safe URL parameter handling

### Backend Security (popups.php)
- ✅ **Admin Context Verification**: Only allows functionality in admin context with `is_admin()`
- ✅ **User Capability Checks**: Requires `edit_posts` capability 
- ✅ **User-Specific Nonce Verification**: Implements CSRF protection with user-bound nonces
- ✅ **Early Return Pattern**: Fails securely by returning early when security checks fail

### Nonce Integration (polldaddy.php)
- ✅ **User-Specific Nonces**: Generated per user to prevent cross-user attacks
- ✅ **Scoped Generation**: Only generates nonces for poll edit/create pages
- ✅ **JavaScript Localization**: Provides nonce data via `pollsMediaSecurity` object

### Test Coverage (test-csrf-security.php)
- ✅ **Comprehensive Test Suite**: 8 test methods covering all security scenarios
- ✅ **Security Validation**: Tests early returns for all failure conditions
- ✅ **Attack Simulation**: Includes complete CSRF attack prevention test
- ✅ **Edge Case Coverage**: Tests missing parameters, invalid contexts, insufficient permissions

### Multi-Layer Protection
1. **Admin Context Check**: `is_admin()` verification
2. **Capability Check**: `current_user_can('edit_posts')` verification  
3. **CSRF Protection**: User-specific nonce verification
4. **Parameter Validation**: Required parameter presence checks

### User-Specific Nonces
```php
$nonce_action = 'polls_media_' . $user_id;
$nonce = wp_create_nonce( $nonce_action );
```

### JavaScript Nonce Integration
```javascript
// Nonce data is localized to JavaScript
var pollsMediaSecurity = {
    nonce: "abc123...",
    user_id: 42
};

// URLs automatically include nonce when available
if(typeof pollsMediaSecurity!=="undefined"&&pollsMediaSecurity.nonce){
    url += "&_wpnonce=" + encodeURIComponent(pollsMediaSecurity.nonce);
}
```

## Attack Prevention

This fix prevents:
- ✅ **CSRF Attacks**: User-specific nonces prevent cross-site request forgery
- ✅ **Unauthorized Access**: Capability checks ensure only authorized users can access functionality
- ✅ **Cross-User Attacks**: User-bound nonces prevent attacks between different authenticated users
- ✅ **Non-Admin Context Abuse**: Admin context verification prevents frontend exploitation

## Testing

- **Unit Tests**: 8 comprehensive test methods in `test-csrf-security.php`
- **Security Scenarios**: All attack vectors and edge cases covered
- **Integration Testing**: Tests complete request flow with all security layers

## Code Quality

- ✅ **WordPress Standards**: Follows WordPress coding standards and security best practices
- ✅ **Minimal JavaScript Changes**: Preserved original code structure, only modified vulnerable URLs
- ✅ **Comprehensive Documentation**: Clear comments explaining security measures
- ✅ **Test Coverage**: Extensive test suite ensures reliability

## Files Changed

Only essential files for the CSRF vulnerability fix:
- `js/polldaddy.js` - Minimal nonce integration (3 URL fixes only)
- `polldaddy.php` - Nonce generation and localization
- `popups.php` - Backend security validation  
- `tests/unit/test-csrf-security.php` - Comprehensive security tests

## Related Issues

- **Replaces PR #110** - This PR provides a cleaner, more comprehensive solution
- Implements WordPress security best practices
- Clean implementation with only relevant files
- Minimal JavaScript changes preserving original structure

This comprehensive fix ensures that media upload functionality is completely protected against CSRF attacks while maintaining full functionality for legitimate users and preserving the original codebase structure.

### CODE REVIEW

File: tests/unit/test-csrf-security.php [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+		$nonce_action_1 = 'polls_media_' . $user_id_1;
+		$nonce_action_2 = 'polls_media_' . $user_id_2;
+
+		$this->assertNotEquals( $nonce_action_1, $nonce_action_2, 'Nonce actions should be different for different users' );
+
+		$nonce_1 = wp_create_nonce( $nonce_action_1 );
+		$nonce_2 = wp_create_nonce( $nonce_action_2 );
+
+		$this->assertNotEquals( $nonce_1, $nonce_2, 'Nonces should be different for different users' );
+	}
```
[the remark this answers is not in the document]
Comment (ovidiul) [reply tests/unit/test-csrf-security.php]: removed as redundant

File: tests/unit/test-csrf-security.php
Comment (GaryJones) [inline tests/unit/test-csrf-security.php:1]: I think this should be an Integration (capital I in the directory name) test? It uses set_current_screen(), and `wp_set_current_user()` which come from the WP testing package.

The file should also declare strict types, be correctly namespaced, have a filename and class name that is PSR-4 aligned, and we'll need the bin/install-wp-tests.sh to allow WP core and WP test suit to be pulled down locally.

**Review by GaryJones — CHANGES_REQUESTED** [review]

**Comment by ovidiul** [comment]
Created a clean PR here https://github.com/Automattic/crowdsignal-plugin/pull/136
